pkg/controller/node/taint_controller.go

```diff
@@ -0,0 +1,394 @@
+		FieldSelector: selector.String(),
+		LabelSelector: labels.Everything().String(),
+	})
+	for i := 0; i < retries && err != nil; i++ {
+		pods, err = c.Core().Pods(v1.NamespaceAll).List(amv1.ListOptions{
+			FieldSelector: selector.String(),
+			LabelSelector: labels.Everything().String(),
+		})
+		time.Sleep(100 * time.Millisecond)
+	}
+	if err != nil {
+		glog.Errorf("Failed to get Pods assigned to node %v. Skipping update.", nodeName)
+	}
+	return pods.Items
+}
+
+func getMinTolerationTime(tolerations []v1.Toleration) time.Duration {
+	minTolerationTime := int64(-1)
+	for i := range tolerations {
+		if tolerations[i].TolerationSeconds != nil {
+			if minTolerationTime < 0 {
+				minTolerationTime = *(tolerations[i].TolerationSeconds)
+			} else {
+				tolerationSeconds := *(tolerations[i].TolerationSeconds)
+				if tolerationSeconds < minTolerationTime {
+					if tolerationSeconds < 0 {
+						minTolerationTime = 0
+					} else {
+						minTolerationTime = tolerationSeconds
+					}
+				}
+			}
+		}
+	}
+	return time.Duration(minTolerationTime) * time.Second
+}
+
+func NewTaintController(c clientset.Interface) *TaintController {
+	return &TaintController{
+		client:             c,
+		taintEvictionQueue: CreateWorkerQueue(deletePodHandler(c)),
+		taintedNodes:       make(map[string][]v1.Taint),
+		nodeUpdateChannel:  make(chan nodeUpdateItem, nodeUpdateChannelSize),
+		podUpdateChannel:   make(chan podUpdateItem, podUpdateChannelSize),
+	}
+}
+
+func (tc *TaintController) Run(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			break
+		case nodeUpdate := <-tc.nodeUpdateChannel:
+			tc.handleNodeUpdate(nodeUpdate)
+		case podUpdate := <-tc.podUpdateChannel:
+		priority:
+			for {
+				select {
+				case nodeUpdate := <-tc.nodeUpdateChannel:
+					tc.handleNodeUpdate(nodeUpdate)
+				default:
+					break priority
+				}
+			}
+			tc.handlePodUpdate(podUpdate)
+		}
+	}
+}
+
+func (tc *TaintController) PodUpdated(oldPod *v1.Pod, newPod *v1.Pod) {
+	var err error
+	oldTolerations := []v1.Toleration{}
+	if oldPod != nil {
+		oldTolerations, err = v1.GetPodTolerations(oldPod)
+		if err != nil {
+			glog.Errorf("Failed to get Tolerations from the old Pod: %v", err)
+			return
+		}
+	}
+	newTolerations := []v1.Toleration{}
+	if newPod != nil {
+		newTolerations, err = v1.GetPodTolerations(newPod)
+		if err != nil {
+			glog.Errorf("Failed to get Tolerations from the new Pod: %v", err)
+			return
+		}
+	}
+
+	if oldPod != nil && newPod != nil && api.Semantic.DeepEqual(oldTolerations, newTolerations) && oldPod.Spec.NodeName == newPod.Spec.NodeName {
+		return
+	}
+	select {
+	case tc.podUpdateChannel <- podUpdateItem{
+		oldPod:         oldPod,
+		newPod:         newPod,
+		oldTolerations: oldTolerations,
+		newTolerations: newTolerations,
+	}:
+	default:
+		// If our update queue is full it means that something's blocked, and we
+		// should try restarting controller.
+		panic(fmt.Errorf("TaintController Pod update queue is full"))
```

Comment:
Yup, but the easiest way to do it is to just restart whole controller. This way it's less code, ergo less error prone. I can try to add something that tries to do resync, but it's possible that we'll leak goroutines there. I'd rather go with panic, because I believe it's safer.